src/test/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandlerTest.java

```diff
@@ -0,0 +1,46 @@
+package org.broadinstitute.barclay.help;
+
+import org.broadinstitute.barclay.help.testinputs.TestArgumentContainer;
+import org.broadinstitute.barclay.help.testinputs.TestExtraDocs;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.Collections;
+
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class DefaultDocWorkUnitHandlerTest {
+
+    private final static DefaultDocWorkUnitHandler DEFAULT_DOC_WORK_UNIT_HANDLER =
+            new DefaultDocWorkUnitHandler(new HelpDoclet());
+
+    @DataProvider
+    public Object[][] workUnitForDocStrings() throws Exception {
+        return new Object[][] {
+                {DocGenMocking.createDocWorkUnit(DEFAULT_DOC_WORK_UNIT_HANDLER, TestArgumentContainer.class),
+                        "", "Argument container class for testing documentation generation.",
+                        "Test feature group name", "Test program group used for testing"},
+                {DocGenMocking.createDocWorkUnit(DEFAULT_DOC_WORK_UNIT_HANDLER, TestExtraDocs.class, DocGenMocking.mockClassDoc("Javadoc description")),
+                        "", "Javadoc description",
+                        "Test extra docs group name", ""},
+                // TODO: because DefaultWorkUnitHandler does not have an extra tag, the mocked class doc uses an empty tag name for the inline tag
+                // TODO: perhaps Barclay should include a default inline tag prefix for being able to re-use between different toolkits
+                {DocGenMocking.createDocWorkUnit(DEFAULT_DOC_WORK_UNIT_HANDLER, TestExtraDocs.class, DocGenMocking.mockClassDoc("Javadoc description \nin two lines", Collections.singletonMap("", "Inline tag"))),
+                        "Inline tag", "Javadoc description in two lines",
+                        "Test extra docs group name", ""}
+        };
+    }
+
+    @Test(dataProvider = "workUnitForDocStrings")
+    public void testGetDocumentationStrings(final DocWorkUnit workUnit,
+            final String description, final String summary,
+            final String groupName, final String groupSummary) throws Exception {
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getDescription(workUnit), description);
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getSummaryForWorkUnit(workUnit), summary);
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getGroupNameForWorkUnit(workUnit), groupName);
+        Assert.assertEquals(DEFAULT_DOC_WORK_UNIT_HANDLER.getGroupSummaryForWorkUnit(workUnit), groupSummary);
```

Comment:
I think we ultimately want to shoot for a structure with (at least) one unit test per method, and symmetry between the methods and test methods, even if we don't have complete unit test coverage now. This test method should be replaced with 4 separate test methods. That would also cut down on the number of "expected" arguments for each test method.

You may need to produce a separate data provider for each test method (like this test, which uses two work units, one from TestArgumentContainer and one from TestExtraDocs, each with the corresponding expected value for the data entry, but thats ok.

Reply:
Also, these unit tests are really helpful to point out flaws in the underlying code - I never fully refactored the code I ported form GATK3 and it shows. All of the methods being tested here are DocWorkUnitHandler methods, but every single one of them takes a workUnit argument, and the implementations only use the work unit itself. These should all be methods on DocWorkUnit! But thats for another day. Still, these kinds of tests are really helpful since they make that easy to see.

Reply:
Refactor to have one test per method.